nats/bench_command.go

```diff
@@ -64,37 +113,123 @@ func (c *benchCmd) bench(_ *kingpin.ParseContext) error {
 	}
 
 	bm := bench.NewBenchmark("NATS", c.numSubs, c.numPubs)
 
 	startwg := &sync.WaitGroup{}
 	donewg := &sync.WaitGroup{}
 
+	var js nats.JetStreamContext
+
+	if (c.request || c.reply) && c.js {
+		log.Fatalf("Request/Reply mode is not applicable to streams")
+	} else if !c.js {
+		if c.request || c.reply {
+			log.Printf("Benchmark in request/reply mode")
+		}
+		if c.request && c.reply {
+			log.Fatalf("Request/Reply mode error: can not be both a requester and a replier at the same time, please use at least two instances of nats bench to benchmark request/reply")
+		}
+		if c.reply && c.numPubs > 0 && c.numSubs > 0 {
+			log.Fatalf("Request/Reply mode error: can not have a publisher while in --reply mode")
+		}
+	}
+
+	if c.js {
+		// create the stream for the benchmark (and purge it)
+		nc, err := nats.Connect(config.ServerURL(), natsOpts()...)
+		if err != nil {
+			log.Fatalf("NATS connection failed: %s", err)
+			return err
+		}
+
+		js, err = nc.JetStream()
+		if err != nil {
+			log.Fatalf("Couldn't create jetstream context: %v", err)
+		}
+
+		// First delete any prior stream unless no delete
+		if !c.noDeleteStream {
+			log.Printf("Deleting any existing stream")
+			js.DeleteStream(JS_STREAM_NAME)
+		}
+
+		storageType := func() nats.StorageType {
+			switch c.storage {
+			case "file":
+				return nats.FileStorage
+			case "memory":
+				return nats.MemoryStorage
+			default:
+				{
+					log.Printf("Unknown storage type %s, using memory", c.storage)
+					return nats.MemoryStorage
+				}
+			}
+		}()
+
+		js.AddStream(&nats.StreamConfig{Name: JS_STREAM_NAME, Subjects: []string{c.subject}, Retention: nats.LimitsPolicy, MaxConsumers: -1, MaxMsgs: -1, MaxBytes: -1, Discard: nats.DiscardOld, MaxAge: 9223372036854775807, MaxMsgsPerSubject: -1, MaxMsgSize: -1, Storage: storageType, Replicas: c.replicas, Duplicates: time.Second * 2})
+		if !c.noPurge {
+			log.Printf("Purging the stream")
+			js.PurgeStream(JS_STREAM_NAME)
+		}
+
+		if !c.noDeleteStream {
+			log.Printf("Will delete the stream at the end of the run")
+			defer js.DeleteStream(JS_STREAM_NAME)
+		}
+
+		// create the pull consumer
+		if c.pull && c.numSubs > 0 {
+			js.AddConsumer(JS_STREAM_NAME, &nats.ConsumerConfig{
+				Durable:       JS_PULLCONSUMER_NAME,
+				DeliverPolicy: nats.DeliverAllPolicy,
+				AckPolicy:     nats.AckExplicitPolicy,
+				ReplayPolicy:  nats.ReplayInstantPolicy,
+				MaxAckPending: c.maxAckPending,
+			})
+			defer js.DeleteConsumer(JS_STREAM_NAME, JS_PULLCONSUMER_NAME)
+		}
+	}
+	subCounts := bench.MsgsPerClient(c.numMsg, c.numSubs)
+
 	for i := 0; i < c.numSubs; i++ {
 		nc, err := nats.Connect(config.ServerURL(), natsOpts()...)
 		if err != nil {
-			return fmt.Errorf("nats connection %d failed: %s", i, err)
+			return fmt.Errorf("NATS connection %d failed: %s", i, err)
```

Comment:
return `nats connection %d failed` as before this will probably fail linting, idiomatic go is to return lower case first word in erorrs.

Reply:
it passes lint but I changed it back (I had capitalized it because I thought it looked better, didn't know about that idiomatic Go aspect), changed back to all lower case now